src/GraphQLParser/AST/Values/GraphQLListValue.cs

```diff
@@ -9,18 +10,53 @@ namespace GraphQLParser.AST;
 [DebuggerDisplay("GraphQLListValue: {Value}")]
 public class GraphQLListValue : GraphQLValue
 {
+    private IList<object?>? _list;
+
     /// <inheritdoc/>
     public override ASTNodeKind Kind => ASTNodeKind.ListValue;
 
     /// <summary>
-    /// List value represented as <see cref="ROM"/>.
+    /// Values of the list represented as a list of nested <see cref="GraphQLValue"/> nodes.
     /// </summary>
-    public ROM Value { get; set; }
+    public List<GraphQLValue>? Values { get; set; }
 
     /// <summary>
-    /// Values of the list represented as a list of nested <see cref="GraphQLValue"/> nodes.
+    /// List value represented as <see cref="List{T}"/>.
+    /// <br/>
+    /// This property allocates the string on the heap on first access
+    /// and then caches it. Call <see cref="Reset"/> to reset cache when needed.
     /// </summary>
-    public List<GraphQLValue>? Values { get; set; }
+    public IList<object?> TypedValue
+    {
+        get
+        {
+            if (_list == null)
+            {
+                if (Values == null || Values.Count == 0)
+                {
+                    _list = Array.Empty<object?>();
+                }
+                else
+                {
+                    var list = new List<object?>(Values.Count);
+                    foreach (var value in Values)
+                        list.Add(value.ClrValue);
+                    _list = list;
```

Comment:
Count is known; this should be an array to eliminate an allocation, especially as arrays implement `IList`.

Reply:
👍  Moved in main repo.